src/lib.rs

```diff
@@ -2230,6 +2230,65 @@ pub trait Itertools : Iterator {
         Ok(start)
     }
 
+    /// Fallibly fold `Result` values from an iterator.
+    ///
+    /// Only `Ok` values are folded. If no error is encountered, the folded
+    /// value is returned inside `Ok`. Otherwise, the operation terminates
+    /// and returns the first `Err` value it encounters. No iterator elements are
+    /// consumed after the first error.
+    ///
+    /// The first accumulator value is the `start` parameter.
+    /// Each iteration passes the accumulator value and the next value inside `Ok`
+    /// to the fold function `f`. If the output is `Err`, then the error is
+    /// returned. Otherwise the value inside the `Ok` variant is extracted and
+    /// used as the next accumulator value.
+    ///
+    /// For example the sequence *Ok(1), Ok(2), Ok(3)* will result in a
+    /// computation like this:
+    ///
+    /// ```ignore
+    /// let mut accum = Ok(start);
+    /// accum = accum.and_then(|val| f(val, 1));
+    /// accum = accum.and_then(|val| f(val, 2));
+    /// accum = accum.and_then(|val| f(val, 3));
+    /// ```
+    ///
+    /// With a `start` value of 0 and an addition as folding function,
+    /// this effectively results in *((0 + 1) + 2) + 3*
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// let values = [1, 2, -2, -1, 2, 1];
+    /// assert_eq!(
+    ///     values.iter()
+    ///           .map(Ok::<_, ()>)
+    ///           .try_fold_ok(0, |sum, &elt| Ok(sum + elt)),
+    ///     Ok(3)
+    /// );
+    /// assert!(
+    ///     values.iter()
+    ///           .map(Ok::<_, &'static str>)
+    ///           .try_fold_ok(0, |sum, &elt| if elt >= 0 { Ok(sum + elt) } else { Err("Negative number") })
+    ///           .is_err()
+    /// );
+    /// assert_eq!(
+    ///     values.iter()
+    ///           .map(|x| if x % 2 == 0 { Ok(x) } else { Err("Odd number") })
+    ///           .try_fold_ok(0, |sum, &elt| if elt >= 0 { Ok(sum + elt) } else { Err("Negative number") }),
+    ///     Err("Odd number")
+    /// );
+    /// ```
+    fn try_fold_ok<A, E, B, F>(&mut self, mut start: B, mut f: F) -> Result<B, E>
+        where Self: Iterator<Item = Result<A, E>>,
+              F: FnMut(B, A) -> Result<B, E>
+    {
+        for result_elt in self {
+            start = result_elt.and_then(|elt| f(start, elt))?;
```

Comment:
Is this the same as
```suggestion
            start = f(start, result_elt?)?;
```
?

Reply:
Since all the error types are the same, yes.

Reply:
I ask mostly because I'm not sure whether adding another itertools method is worth it, if the difference is between 

```rust
.try_fold_results(init, |a, x| a.blah(x))
```
and

```rust
.try_fold(init, |a, x| a.blah(x?))
```

Is there an ergonomic reason that it'd be worth having the extra method anyway?

Reply:
No, I suppose once you throw `?` in there's no good reason for the `try_` variants. It really is `.try_fold(init, |a, x| f(a, x?))`.